File: frontend/src/stores/members.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { WorkspaceMember, Invitation, InvitationInfo, MyMembershipResponse, AcceptInviteResponse, MemberRole } from '../types'
import { get, post, patch, del } from './client'

export const useMembersStore = defineStore('members', () => {
  const members = ref<WorkspaceMember[]>([])
  const invitations = ref<Invitation[]>([])
  const myMembership = ref<MyMembershipResponse | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchMembers(workspaceId: number) {
    loading.value = true
    error.value = null
    try {
      members.value = await get<WorkspaceMember[]>(`/workspaces/${workspaceId}/members`)
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  async function fetchInvitations(workspaceId: number) {
    loading.value = true
    error.value = null
    try {
      const items = await get<Invitation[]>(`/workspaces/${workspaceId}/members/invitations`)
      invitations.value = items
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  async function invite(workspaceId: number, email: string, role: MemberRole) {
    error.value = null
    try {
      const inv = await post<Invitation>(`/workspaces/${workspaceId}/members/invite`, { email, role })
      invitations.value.push(inv)
      return inv
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  async function removeMember(workspaceId: number, userId: number) {
    error.value = null
    try {
      await del(`/workspaces/${workspaceId}/members/${userId}`)
      members.value = members.value.filter((m) => m.user_id !== userId)
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  async function updateRole(workspaceId: number, userId: number, role: MemberRole) {
    error.value = null
    try {
      const updated = await patch<WorkspaceMember>(`/workspaces/${workspaceId}/members/${userId}/role`, { role })
      const idx = members.value.findIndex((m) => m.user_id === userId)
      if (idx !== -1) members.value[idx] = updated
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  async function revokeInvitation(workspaceId: number, invitationId: number) {
    error.value = null
    try {
      await del(`/workspaces/${workspaceId}/members/invitations/${invitationId}`)
      invitations.value = invitations.value.filter((i) => i.id !== invitationId)
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  async function fetchMyMembership(workspaceId: number) {
    try {
      myMembership.value = await get<MyMembershipResponse>(`/workspaces/${workspaceId}/members/me`)
      return myMembership.value
    } catch {
      myMembership.value = null
      return null
    }
  }

  async function acceptInvite(token: string) {
    error.value = null
    try {
      const res = await post<AcceptInviteResponse>(`/invitations/${token}/accept`)
      return res
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  async function getInviteInfo(token: string) {
    try {
      return await get<InvitationInfo>(`/invitations/${token}`)
    } catch (e) {
      error.value = (e as Error).message
      throw e
    }
  }

  return {
    members,
    invitations,
    myMembership,
    loading,
    error,
    fetchMembers,
    fetchInvitations,
    fetchMyMembership,
    invite,
    removeMember,
    updateRole,
    revokeInvitation,
    acceptInvite,
    getInviteInfo,
  }
})

File: routes/workspace/members.py
from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select

import models
from models.collaboration import InvitationStatus
from schemas.workspace.collaboration import (
    AcceptResponse,
    InvitationInfoResponse,
    InvitationResponse,
    InviteRequest,
    MemberResponse,
    RoleUpdateRequest,
)
from services.auth import CurrentUser
from services.workspace.collaboration import CollaborationService
from utils.email import send_invite_email
from services.workspace.workspaces import WorkspaceService

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/api/workspaces/{workspace_id}/members",
    tags=["Members"],
)


async def _get_workspace_or_404(
    workspace_id: int,
    ws_service: WorkspaceService,
) -> models.Workspace:
    workspace = await ws_service.get_by_id(workspace_id)
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Workspace not found"
        )
    return workspace


@router.get("", response_model=list[MemberResponse])
async def list_members(
    workspace_id: int,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    await _get_workspace_or_404(workspace_id, ws_service)
    await collab.require_access(workspace_id, current_user.id, min_role="member")
    return await collab.list_members(workspace_id)


@router.post("/invite", response_model=InvitationResponse, status_code=status.HTTP_201_CREATED)
async def invite_member(
    workspace_id: int,
    body: InviteRequest,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    workspace = await _get_workspace_or_404(workspace_id, ws_service)
    await collab.require_access(workspace_id, current_user.id, min_role="admin")

    if body.role not in ("admin", "member"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Role must be 'admin' or 'member'",
        )

    invitation = await collab.create_invite(
        workspace_id=workspace_id,
        email=body.email,
        role=body.role,
        invited_by_id=current_user.id,
    )

    try:
        await send_invite_email(
            to_email=invitation.email,
            token=invitation.token,
            role=invitation.role.value,
            workspace_title=workspace.title,
            invited_by_username=current_user.username,
        )
    except Exception as exc:
        logger.warning("Invite email failed but invitation was created: %s", exc)

    return InvitationResponse.model_validate(invitation)


@router.get("/invitations", response_model=list[InvitationResponse])
async def list_invitations(
    workspace_id: int,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    await _get_workspace_or_404(workspace_id, ws_service)
    await collab.require_access(workspace_id, current_user.id, min_role="admin")
    invitations = await collab.list_invitations(workspace_id)
    return [InvitationResponse.model_validate(inv) for inv in invitations]


@router.delete("/invitations/{invitation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def revoke_invitation(
    workspace_id: int,
    invitation_id: int,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    await _get_workspace_or_404(workspace_id, ws_service)
    await collab.require_access(workspace_id, current_user.id, min_role="admin")

    result = await collab.db.execute(
        select(models.Invitation).where(
            models.Invitation.id == invitation_id,
            models.Invitation.workspace_id == workspace_id,
        )
    )
    invitation = result.scalars().first()
    if not invitation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Invitation not found"
        )

    invitation.status = InvitationStatus.revoked
    await collab.db.commit()


@router.patch("/{user_id}/role", response_model=MemberResponse)
async def update_member_role(
    workspace_id: int,
    user_id: int,
    body: RoleUpdateRequest,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    await _get_workspace_or_404(workspace_id, ws_service)
    member = await collab.update_role(
        workspace_id=workspace_id,
        user_id=user_id,
        new_role=body.role,
        by_user_id=current_user.id,
    )
    return MemberResponse.model_validate(member)


@router.delete("/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_member(
    workspace_id: int,
    user_id: int,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    await _get_workspace_or_404(workspace_id, ws_service)
    await collab.remove_member(
        workspace_id=workspace_id,
        user_id=user_id,
        by_user_id=current_user.id,
    )


invitations_router = APIRouter(
    prefix="/api/invitations",
    tags=["Invitations"],
)


@invitations_router.get("/{token}", response_model=InvitationInfoResponse)
async def get_invitation_info(
    token: str,
    collab: Annotated[CollaborationService, Depends()],
    ws_service: Annotated[WorkspaceService, Depends()],
):
    result = await collab.db.execute(
        select(models.Invitation).where(models.Invitation.token == token)
    )
    invitation = result.scalars().first()
    if not invitation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Invitation not found"
        )

    workspace = await ws_service.get_by_id(invitation.workspace_id)
    return InvitationInfoResponse(
        id=invitation.id,
        email=invitation.email,
        role=invitation.role.value,
        status=invitation.status.value,
        expires_at=invitation.expires_at,
        workspace={"id": workspace.id if workspace else None, "title": workspace.title if workspace else None},
    )


@invitations_router.post("/{token}/accept", response_model=AcceptResponse, status_code=status.HTTP_200_OK)
async def accept_invitation(
    token: str,
    current_user: CurrentUser,
    collab: Annotated[CollaborationService, Depends()],
):
    member = await collab.accept_invite(token=token, user_id=current_user.id)
    return AcceptResponse.model_validate(member)
